File: space-game/Core/Classes/Range.vb
Namespace Classes
    Public Class Range

        Private ReadOnly _values As New List(Of Double)
        Private _calculated As Boolean

        Public Sub New(defaultLowerBound As Double, defaultUpperBound As Double)
            If defaultLowerBound > defaultUpperBound Then
                Throw New ArgumentOutOfRangeException(String.Format("{0} should be greater than {1}", NameOf(defaultUpperBound), NameOf(defaultLowerBound)))
            End If
            _lowerBound = defaultLowerBound
            _upperBound = defaultUpperBound
            _calculated = True
        End Sub

        Public Sub AddValue(value As Double)
            _values.Add(value)
            _calculated = False
        End Sub

        Private _lowerBound As Double
        Public ReadOnly Property LowerBound As Double
            Get
                Return GetLowerBound()
            End Get
        End Property

        Protected Function GetLowerBound() As Double
            If Not _calculated Then
                CalculateValues()
            End If
            Return _lowerBound
        End Function

        Private _upperBound As Double
        Public ReadOnly Property UpperBound As Double
            Get
                Return GetUpperBound()
            End Get
        End Property

        Protected Function GetUpperBound() As Double
            If Not _calculated Then
                CalculateValues()
            End If
            Return _upperBound
        End Function

        Private Sub CalculateValues()
            _values.Sort()
            _lowerBound = _values(0)
            _upperBound = _values(_values.Count - 1)
            _calculated = True
        End Sub
    End Class
End Namespace

File: space-game/Core/Classes/Range.cs
using System;
using System.Collections.Generic;

namespace Core.Classes
{
    public class Range
    {

        private readonly List<double> _values = new List<double>();

        private bool _calculated;
        public Range(double defaultLowerBound, double defaultUpperBound)
        {
            if (defaultLowerBound > defaultUpperBound)
            {
                throw new ArgumentOutOfRangeException($"{nameof(defaultUpperBound)} should be greater than {nameof(defaultLowerBound)}");
            }
            _lowerBound = defaultLowerBound;
            _upperBound = defaultUpperBound;
            _calculated = true;
        }

        public void AddValue(double value)
        {
            _values.Add(value);
            _calculated = false;
        }

        private double _lowerBound;
        public double LowerBound => GetLowerBound();

        protected double GetLowerBound()
        {
            if (!_calculated)
            {
                CalculateValues();
            }
            return _lowerBound;
        }

        private double _upperBound;
        public double UpperBound => GetUpperBound();

        protected double GetUpperBound()
        {
            if (!_calculated)
            {
                CalculateValues();
            }
            return _upperBound;
        }

        private void CalculateValues()
        {
            _values.Sort();
            _lowerBound = _values[0];
            _upperBound = _values[_values.Count - 1];
            _calculated = true;
        }
    }
}
